### PR TITLE
1077 UI changes for stocktake reasons 

### LINKED ISSUE #1077: UI changes for Stocktake reasons
Continuation of #870

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #1077 

- Enable reasons editing if adjustment reasons is active
- Indicate required lines with an asterisk 
- Enable error box on error

### CODE REVIEW

File: client/packages/system/src/InventoryAdjustmentReason/Components/InventoryAdjustmentReasonSearchInput.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+          <BasicTextInput
+            {...props}
+            autoFocus={autoFocus}
+            InputProps={{
+              disableUnderline: false,
+              style: props.disabled ? { paddingLeft: 0 } : {},
+              ...props.InputProps,
+            }}
+            sx={{ width }}
+            error={isError}
```
Comment (roxy-dao) [inline client/packages/system/src/InventoryAdjustmentReason/Components/InventoryAdjustmentReasonSearchInput.tsx:67]: The batch api doesn't return error based on line, so all the boxes are red on error 🙈

**Comment by mark-prins** [comment]
I wanted to comment on a couple of things, seemed easier to experiment, so I have a suggestion here: https://github.com/openmsupply/open-msupply/compare/1077-UI-changes-for-Stocktake-reasons-...1077-review-suggestions

- When you click ok on the modal, there are two error snacks shown, would be nice to have one
- and have it more specific
- The autocomplete is throwing a warning in console
- The datatable Viewport implementation is giving an error in console
- The date pickers are complaining - was updated in programs, but I'm tired of waiting for that to come in
- Would be good to only highlight the inputs which require a reason when there's an error

File: client/packages/system/src/InventoryAdjustmentReason/Components/InventoryAdjustmentReasonSearchInput.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  onChange,
+  autoFocus = false,
+  stockReduction,
+  isError,
+}) => {
+  const { data, isLoading } =
+    useInventoryAdjustmentReason.document.listAllActive();
+  const disabled = data?.totalCount === 0;
+
+  return (
```
Comment (mark-prins) [inline client/packages/system/src/InventoryAdjustmentReason/Components/InventoryAdjustmentReasonSearchInput.tsx:40]: In my datafile I have some confusing names for the reasons, which led to errors when trying to save ( reducing stock and using a positive reason )

Screenshot 2023-02-01 at 11 31 08 AM

Screenshot 2023-02-01 at 11 38 11 AM

I've added `type` to the query, and used it to filter the reasons:

```
const reasonFilter = (reason: InventoryAdjustmentReasonRowFragment) => {
    if (stockReduction === 0 || !stockReduction) return true;
    if (stockReduction < 0)
      return reason.type === InventoryAdjustmentReasonNodeType.Positive;
    return reason.type === InventoryAdjustmentReasonNodeType.Negative;
  };
  const reasons = (data?.nodes ?? []).filter(reasonFilter);
  const options = defaultOptionMapper(reasons, 'reason');


/// then

        options={options}

```

**Review by mark-prins — CHANGES_REQUESTED** [review]
👍 That's so good!

I have some suggestions - the main issue would be filtering by type, so that we only show 'NEGATIVE' if you are reducing stock. the others were window dressing

I also wonder if the 'difference' column is the wrong way around - if I enter a # of packs more than the snapshot, the difference is negative, and that doesn't make sense to me 🤔 

Screenshot 2023-02-01 at 11 33 15 AM

File: client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts [hunk 3]
```diff
@@ -255,6 +256,44 @@ export const useStocktakeColumns = ({
           }
         },
       },
+      {
```
Comment (mark-prins) [inline client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts:259]: would you mind changing the `let` above to `const` the variable `total` isn't reassigned, so doesn't need to be a `let` (lines 223 & 242)

Comment (roxy-dao) [reply client/packages/inventory/src/Stocktake/DetailView/ContentArea/columns.ts:279]: [19b9b54](https://github.com/openmsupply/open-msupply/pull/1098/commits/19b9b547419b49259beb34ac375b1f3046d4cd96) all changes (:

**Review by mark-prins — APPROVED** [review]
🙏  looking good

**Pull request merged**